File: app/src/main/java/farayan/sabad/ui/HomeFragment.java
package farayan.sabad.ui;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.google.zxing.client.android.BeepManager;

import java.util.Objects;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import farayan.commons.Commons.RecyclerViewOrientations;
import farayan.commons.Commons.Rial;
import farayan.commons.FarayanUtility;
import farayan.commons.QueryBuilderCore.ComparableFilter;
import farayan.commons.QueryBuilderCore.SortConfig;
import farayan.commons.QueryBuilderCore.SortDirections;
import farayan.commons.QueryBuilderCore.TextFilter;
import farayan.commons.QueryBuilderCore.TextMatchModes;
import farayan.sabad.R;
import farayan.sabad.SabadConstants;
import farayan.sabad.SabadUtility;
import farayan.sabad.constants.SabadFragmentEvents;
import farayan.sabad.core.OnePlace.CategoryGroup.ICategoryGroupRepo;
import farayan.sabad.core.OnePlace.Group.GroupEntity;
import farayan.sabad.core.OnePlace.Group.GroupParams;
import farayan.sabad.core.OnePlace.Group.GroupSchema;
import farayan.sabad.core.OnePlace.Group.GroupUniqueNameNeededException;
import farayan.sabad.core.OnePlace.Group.IGroupRepo;
import farayan.sabad.core.OnePlace.Group.NewGroupNameNeededException;
import farayan.sabad.core.OnePlace.GroupUnit.IGroupUnitRepo;
import farayan.sabad.core.OnePlace.Invoice.IInvoiceRepo;
import farayan.sabad.core.OnePlace.InvoiceItem.IInvoiceItemRepo;
import farayan.sabad.core.OnePlace.ProductBarcode.IProductBarcodeRepo;
import farayan.sabad.core.OnePlace.Store.IStoreRepo;
import farayan.sabad.core.OnePlace.StoreCategory.IStoreCategoryRepo;
import farayan.sabad.core.OnePlace.StoreGroup.IStoreGroupRepo;
import farayan.sabad.core.OnePlace.Unit.IUnitRepo;
import farayan.sabad.core.model.product.IProductRepo;
import farayan.sabad.models.Group.GroupRecyclerAdapter;
import farayan.sabad.vms.InvoiceItemFormViewModel;


@AndroidEntryPoint
public class HomeFragment extends HomeFragmentParent
{
	private InvoiceItemFormViewModel invoiceItemFormViewModel ;//= new ViewModelProvider(requireActivity()).get(InvoiceItemFormViewModel.class);
	private final View.OnClickListener EditButtonOnClickListener = new View.OnClickListener()
	{
		@Override
		public void onClick(View view) {
			String name = Objects.requireNonNull(QueryEditText().getText()).toString();
			if (FarayanUtility.IsNullOrEmpty(name))
				return;
			GroupParams groupParams = new GroupParams();
			groupParams.QueryableName = new TextFilter(FarayanUtility.Queryable(name));
			GroupEntity groupEntity = TheGroupRepo.First(groupParams);
			if (groupEntity == null) {
				return;
			}
			GroupFormDialog dialog = new GroupFormDialog(
					new GroupFormDialog.Input(
							groupEntity,
							TheGroupRepo,
							TheInvoiceItemRepo,
							TheProductRepo,
							TheGroupUnitRepo,
							x -> Reload(),
							x -> Reload()
					),
					requireActivity()
			);
			dialog.show();
		}
	};
	private boolean reloaded;

	@Inject
	public void Inject(
			IStoreCategoryRepo storeCategoryRepo,
			IStoreGroupRepo storeGroupRepo,
			ICategoryGroupRepo categoryGroupRepo,
			IGroupRepo GroupRepo,
			IStoreRepo storeRepo,
			IInvoiceRepo invoiceRepo,
			IInvoiceItemRepo invoiceItemRepo,
			IGroupUnitRepo GroupUnitRepo,
			IProductRepo productRepo,
			IProductBarcodeRepo productBarcodeRepo,
			IUnitRepo unitRepo
	) {
		TheStoreCategoryRepo = storeCategoryRepo;
		TheStoreGroupRepo = storeGroupRepo;
		TheCategoryGroupRepo = categoryGroupRepo;
		TheGroupRepo = GroupRepo;
		TheStoreRepo = storeRepo;
		TheInvoiceRepo = invoiceRepo;
		TheInvoiceItemRepo = invoiceItemRepo;
		TheGroupUnitRepo = GroupUnitRepo;
		TheProductRepo = productRepo;
		TheProductBarcodeRepo = productBarcodeRepo;
		TheUnitRepo = unitRepo;
		init();
	}

	private IStoreCategoryRepo TheStoreCategoryRepo;
	private IStoreGroupRepo TheStoreGroupRepo;
	private ICategoryGroupRepo TheCategoryGroupRepo;
	private IGroupRepo TheGroupRepo;
	private IStoreRepo TheStoreRepo;
	private IInvoiceRepo TheInvoiceRepo;
	private IInvoiceItemRepo TheInvoiceItemRepo;
	private IGroupUnitRepo TheGroupUnitRepo;
	private IProductRepo TheProductRepo;
	private IProductBarcodeRepo TheProductBarcodeRepo;
	private IUnitRepo TheUnitRepo;

	@Override
	public int TitleID() {
		return R.string.home_title;
	}

	@SuppressLint("NonConstantResourceId")
	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		switch (item.getItemId()) {
			case R.id.ScanMenu:
				StartScan();
				break;
			case R.id.CheckoutMenu:
				TryCheckout();
				break;
			case R.id.DisplayInvoicesMenu:
				HostActivity.OnFragmentCalled(null, SabadFragmentEvents.DisplayInvoices);
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	public enum Requests
	{
		Scan(1);

		public final int ID;

		Requests(int id) {
			ID = id;
		}
	}

	private void StartScan() {
		SabadUtility.BarcodeScanCameraPermission(
				HostActivity,
				() -> {
					Intent intent = new Intent(getActivity(), ScanActivity.class);
					startActivityForResult(intent, Requests.Scan.ID);
				},
				false
		);
	}

	private void TryCheckout() {

		GroupParams remainedGroupsParams = new GroupParams();
		remainedGroupsParams.Needed = new ComparableFilter<>(true);
		remainedGroupsParams.Picked = new ComparableFilter<>(true);
		remainedGroupsParams.Deleted = new ComparableFilter<>(false);
		long pickedCount = TheGroupRepo.Count(remainedGroupsParams);

		if (pickedCount == 0) {
			FarayanUtility.ShowToastFormatted(getActivity(), "برای ثبت صورتحساب و صندوق، دستکم یکی از نیازمندی‌ها را بردارید");
			return;
		}

		DisplayCheckout();
	}

	private void AutoCheckout() {
		GroupParams neededGroupsParams = new GroupParams();
		neededGroupsParams.Needed = new ComparableFilter<>(true);
		long neededCount = TheGroupRepo.Count(neededGroupsParams);

		GroupParams remainedGroupsParams = new GroupParams();
		remainedGroupsParams.Needed = new ComparableFilter<>(true);
		remainedGroupsParams.Picked = new ComparableFilter<>(false);
		long remainedCount = TheGroupRepo.Count(remainedGroupsParams);

		if (neededCount > 0 && remainedCount == 0) {
			if (SabadUtility.DisplayCheckoutDialogAutomatically(getActivity())) {
				if (SabadUtility.CheckoutGuideNeeded(getActivity()))
					FarayanUtility.ShowToastFormatted(getActivity(), false, "چون تمام کالاهای فهرست خرید (روشن) را برداشنه‌اید، می‌توانید پایان خرید را بزنید");
				handler.postDelayed(this::DisplayCheckout, 5000);
			}
		}
	}

	private void DisplayCheckout() {
		CheckoutPurchaseDialog dialog = new CheckoutPurchaseDialog(
				new CheckoutPurchaseDialog.Input(
						TheStoreCategoryRepo,
						TheStoreGroupRepo,
						TheCategoryGroupRepo,
						TheGroupRepo,
						TheStoreRepo,
						TheInvoiceRepo,
						TheInvoiceItemRepo,
						invoiceEntity -> {
							Reload();
							HostActivity.OnFragmentCalled(null, SabadFragmentEvents.DisplayInvoices);
						}
				),
				HostActivity
		);

		dialog.show();
	}

	private final TextView.OnEditorActionListener QueryEditTextOnEditorActionListener = (v, actionId, event) -> {
		if (FarayanUtility.IsNullOrEmpty(QueryEditText().getText()))
			return false;
		String query = Objects.requireNonNull(QueryEditText().getText()).toString();
		if (actionId != EditorInfo.IME_ACTION_DONE)
			return false;
		if (Persist(false))
			return true;

		GroupEntity groupEntity = TheGroupRepo.FirstByName(query);

		if (QueryEditText().getSelectionStart() != 0 || QueryEditText().getSelectionEnd() != query.length()) {
			String message = groupEntity.Needed ? "«%s» هم‌اکنون در  فهرست خرید است، برای ثبت کالای برداشته‌شده، یکبار دیگر اینتر صفحه‌کلید را بزنید" : "«%s» هم‌اکنون موجود است، برای افزودن به فهرست خرید، یکبار دیگر اینتر صفحه‌کلید را بزنید";
			FarayanUtility.ShowToastFormatted(
					getActivity(),
					message,
					query
			);
			QueryEditText().selectAll();
			return true;
		}
		if (!groupEntity.Needed) {
			FarayanUtility.ShowToastFormatted(
					getActivity(),
					"«%s» به فهرست خرید افزوده شد، برای ثبت کالای برداشته‌شده، یکبار دیگر اینتر صفحه‌کلید را بزنید",
					query
			);

			groupEntity.Needed = true;
			TheGroupRepo.Update(groupEntity);
			Reload();
			return true;
		}

		if (!groupEntity.Picked) {
			InvoiceItemFormDialog dialog = new InvoiceItemFormDialog(
					new InvoiceItemFormInputArgs(
							groupEntity.Item,
							groupEntity,
							null,
							null,
							invoiceItemEntity -> Reload(),
							invoiceItemEntity -> Reload(),
							TheGroupRepo,
							TheGroupUnitRepo,
							TheProductRepo,
							TheProductBarcodeRepo,
							TheInvoiceItemRepo,
							TheUnitRepo,
							new BeepManager(requireActivity())
					),
					(AppCompatActivity) requireActivity(),
					true,
					null,
					invoiceItemFormViewModel
			);
			dialog.show();
			return true;
		}
		return false;
	};

	private final Runnable ReloadCallback = () -> Reload(Objects.requireNonNull(QueryEditText().getText()).toString());
	private final TextWatcher QueryTextChangedListener = new TextWatcher()
	{
		@Override
		public void beforeTextChanged(CharSequence s, int start, int count, int after) {

		}

		@Override
		public void onTextChanged(CharSequence s, int start, int before, int count) {
			handler.removeCallbacks(ReloadCallback);
			handler.postDelayed(ReloadCallback, 500);
		}

		@Override
		public void afterTextChanged(Editable s) {
			QueryLayout().setError(null);
		}
	};
	private final View.OnClickListener RegisterButtonOnClickListener = v -> Persist(true);

	private boolean Persist(boolean toast) {
		try {
			String name = Objects.requireNonNull(QueryEditText().getText()).toString();
			TheGroupRepo.New(name, null, null, true);
			Reload(name);
			QueryLayout().setError(null);
			if (SabadUtility.NewGroupGuideNeeded(getActivity()))
				FarayanUtility.ShowToastFormatted(getActivity(), false, getString(R.string.NewGroupGuide));
			return true;
		} catch (NewGroupNameNeededException e) {
			QueryLayout().setError(getString(R.string.NameNeeded));
			if (toast)
				FarayanUtility.ShowToast(getActivity(), getString(R.string.NameNeeded));
		} catch (GroupUniqueNameNeededException e) {
			QueryLayout().setError(getString(R.string.CategoryUsedName));
			if (toast)
				FarayanUtility.ShowToast(getActivity(), getString(R.string.CategoryUsedName));
		}
		return false;
	}

	@Override
	public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
		inflater.inflate(R.menu.home_menu, menu);
	}


	Handler handler = new Handler();

	protected void InitializeLayout() {
		FarayanUtility.prepareRecyclerView(
				GroupsRecyclerView(),
				RecyclerViewOrientations.Vertical,
				null,
				true,
				true
		);
		init();
		QueryEditText().addTextChangedListener(QueryTextChangedListener);
		EditButton().setOnClickListener(EditButtonOnClickListener);
		RegisterButton().setOnClickListener(RegisterButtonOnClickListener);
		QueryEditText().setOnEditorActionListener(QueryEditTextOnEditorActionListener);
	}

	private void Reload(String query) {
		query = FarayanUtility.Queryable(query);
		GroupParams groupParams = new GroupParams();
		groupParams.Deleted = new ComparableFilter<>(false);
		groupParams.QueryableName = new TextFilter(query);
		groupParams.Sorts.add(new SortConfig(GroupSchema.Needed, SortDirections.Descending));
		groupParams.Sorts.add(new SortConfig(GroupSchema.Picked, SortDirections.Ascending));
		groupParams.Sorts.add(new SortConfig(GroupSchema.Category, SortDirections.Ascending));
		groupParams.Sorts.add(new SortConfig(GroupSchema.PurchasedCount, SortDirections.Descending));
		groupParams.Sorts.add(new SortConfig(GroupSchema.Importance, SortDirections.Descending));

		GroupRecyclerAdapter theAdapter = new GroupRecyclerAdapter(
				getActivity(),
				TheGroupRepo.All(groupParams),
				GroupEntity -> {
					PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient);
					AutoCheckout();
				},
				GroupEntity -> {
					PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient);
					AutoCheckout();
				},
				component -> {
					InvoiceItemFormDialog dialog = new InvoiceItemFormDialog(
							new InvoiceItemFormInputArgs(
									component.getEntity().Item,
									component.getEntity(),
									null,
									null,
									invoiceItemEntity -> {
										component.DisplayEntity(component.getEntity().RefreshForced(TheGroupRepo));
										PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient);
										AutoCheckout();
									},
									invoiceItemEntity -> {
										component.DisplayEntity(component.getEntity().RefreshForced(TheGroupRepo));
										PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient);
									},
									TheGroupRepo,
									TheGroupUnitRepo,
									TheProductRepo,
									TheProductBarcodeRepo,
									TheInvoiceItemRepo,
									TheUnitRepo,
									new BeepManager(requireActivity())
							),
							(AppCompatActivity) requireActivity(),
							true,
							null,
							invoiceItemFormViewModel
					);
					dialog.show();
				},
				GroupEntity -> Reload(),
				Rial.Coefficients.Rial
		);
		GroupsRecyclerView().setAdapter(theAdapter);

		groupParams.QueryableName = new TextFilter(query, TextMatchModes.Exactly);
		boolean found = TheGroupRepo.Count(groupParams) == 1;
		RegisterButton().setVisibility(found ? View.GONE : View.VISIBLE);
		EditButton().setVisibility(found ? View.VISIBLE : View.GONE);
		RegisterButton().setEnabled(FarayanUtility.IsUsable(query));
		PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient);
	}

	private void init() {
		if (reloaded)
			return;
		if (TheGroupRepo == null)
			return;
		if (QueryEditText() == null)
			return;
		Reload(Objects.requireNonNull(QueryEditText().getText()).toString());
		reloaded = true;
	}

	private void Reload() {
		Reload(Objects.requireNonNull(QueryEditText().getText()).toString());
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
		if (requestCode == Requests.Scan.ID && resultCode == SabadConstants.DataChangedResultCode) {
			Reload();
		} else {
			super.onActivityResult(requestCode, resultCode, data);
		}
	}

	@Override
	public boolean HandleBackPressed() {
		if (FarayanUtility.IsUsable(QueryEditText().getText())) {
			QueryEditText().setText("");
			return true;
		}
		return super.HandleBackPressed();
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		invoiceItemFormViewModel = new ViewModelProvider(requireActivity()).get(InvoiceItemFormViewModel.class);
	}
}

File: app/src/main/java/farayan/sabad/ui/HomeFragment.kt
@file:Suppress("UnusedImport")

package farayan.sabad.ui

//import androidx.compose.foundation.lazy.items
import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalLayoutDirection
import androidx.compose.ui.unit.LayoutDirection
import androidx.lifecycle.compose.LocalLifecycleOwner
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dagger.hilt.android.AndroidEntryPoint
import farayan.commons.Commons.Rial
import farayan.commons.FarayanUtility
import farayan.commons.QueryBuilderCore.ComparableFilter
import farayan.commons.QueryBuilderCore.SortConfig
import farayan.commons.QueryBuilderCore.SortDirections
import farayan.commons.QueryBuilderCore.TextFilter
import farayan.commons.QueryBuilderCore.TextMatchModes
import farayan.sabad.R
import farayan.sabad.SabadConstants
import farayan.sabad.SabadUtility
import farayan.sabad.composables.CategoriesCategoryRemoteItemComposable
import farayan.sabad.composables.CategoriesCategoryWithItemsComposable
import farayan.sabad.constants.SabadFragmentEvents
import farayan.sabad.core.OnePlace.CategoryGroup.ICategoryGroupRepo
import farayan.sabad.core.OnePlace.Group.GroupEntity
import farayan.sabad.core.OnePlace.Group.GroupParams
import farayan.sabad.core.OnePlace.Group.GroupSchema
import farayan.sabad.core.OnePlace.Group.GroupUniqueNameNeededException
import farayan.sabad.core.OnePlace.Group.IGroupRepo
import farayan.sabad.core.OnePlace.Group.NewGroupNameNeededException
import farayan.sabad.core.OnePlace.GroupUnit.IGroupUnitRepo
import farayan.sabad.core.OnePlace.Invoice.IInvoiceRepo
import farayan.sabad.core.OnePlace.Invoice.InvoiceEntity
import farayan.sabad.core.OnePlace.Store.IStoreRepo
import farayan.sabad.core.OnePlace.StoreCategory.IStoreCategoryRepo
import farayan.sabad.core.OnePlace.StoreGroup.IStoreGroupRepo
import farayan.sabad.core.OnePlace.Unit.IUnitRepo
import farayan.sabad.core.model.product.IProductRepo
import farayan.sabad.db.Category
import farayan.sabad.db.Item
import farayan.sabad.db.Product
import farayan.sabad.models.Group.GroupRecyclerAdapter
import farayan.sabad.repo.CategoryRepo
import farayan.sabad.utility.hasValue
import farayan.sabad.vms.HomeViewModel
import farayan.sabad.vms.InvoiceItemFormViewModel
import farayan.sabad.vms.InvoiceItemFormViewModel.Companion.Factory
import java.util.Objects
import javax.inject.Inject

@AndroidEntryPoint
class HomeFragment : HomeFragmentParent() {
    var handler: Handler = Handler()
    private var invoiceItemFormViewModel: InvoiceItemFormViewModel? = null //= new ViewModelProvider(requireActivity()).get(InvoiceItemFormViewModel.class);
    private lateinit var homeViewModel: HomeViewModel
    private var reloaded = false
    private var TheStoreCategoryRepo: IStoreCategoryRepo? = null
    private var TheStoreGroupRepo: IStoreGroupRepo? = null
    private var TheCategoryGroupRepo: ICategoryGroupRepo? = null
    private var TheGroupRepo: IGroupRepo? = null
    private var TheStoreRepo: IStoreRepo? = null
    private var TheInvoiceRepo: IInvoiceRepo? = null
    private var TheGroupUnitRepo: IGroupUnitRepo? = null
    private var TheProductRepo: IProductRepo? = null
    private var TheUnitRepo: IUnitRepo? = null
    private var categoryRepo: CategoryRepo? = null

    private val EditButtonOnClickListener = View.OnClickListener {
        val name = Objects.requireNonNull(QueryEditText().text).toString()
        if (FarayanUtility.IsNullOrEmpty(name)) return@OnClickListener
        val groupParams = GroupParams()
        groupParams.QueryableName = TextFilter(FarayanUtility.Queryable(name))
        val groupEntity = TheGroupRepo!!.First(groupParams) ?: return@OnClickListener
        val dialog = GroupFormDialog(
            GroupFormDialog.Input(
                groupEntity,
                TheGroupRepo!!,
                TheProductRepo!!,
                TheGroupUnitRepo!!,
                { x: GroupEntity? -> Reload() },
                { x: GroupEntity? -> Reload() }
            ),
            requireActivity()
        )
        dialog.show()
    }
    private val QueryEditTextOnEditorActionListener = OnEditorActionListener { v: TextView?, actionId: Int, event: KeyEvent? ->
        if (FarayanUtility.IsNullOrEmpty(QueryEditText().text)) return@OnEditorActionListener false
        val query = Objects.requireNonNull(QueryEditText().text).toString()
        if (actionId != EditorInfo.IME_ACTION_DONE) return@OnEditorActionListener false
        if (Persist(false)) return@OnEditorActionListener true

        val groupEntity = TheGroupRepo!!.FirstByName(query)

        if (QueryEditText().selectionStart != 0 || QueryEditText().selectionEnd != query.length) {
            val message =
                if (groupEntity!!.Needed) "«%s» هم‌اکنون در  فهرست خرید است، برای ثبت کالای برداشته‌شده، یکبار دیگر اینتر صفحه‌کلید را بزنید" else "«%s» هم‌اکنون موجود است، برای افزودن به فهرست خرید، یکبار دیگر اینتر صفحه‌کلید را بزنید"
            FarayanUtility.ShowToastFormatted(
                activity,
                message,
                query
            )
            QueryEditText().selectAll()
            return@OnEditorActionListener true
        }
        if (!groupEntity!!.Needed) {
            FarayanUtility.ShowToastFormatted(
                activity,
                "«%s» به فهرست خرید افزوده شد، برای ثبت کالای برداشته‌شده، یکبار دیگر اینتر صفحه‌کلید را بزنید",
                query
            )

            groupEntity.Needed = true
            TheGroupRepo!!.Update(groupEntity)
            Reload()
            return@OnEditorActionListener true
        }

        if (!groupEntity.Picked) {
            //invoiceItemFormViewModel!!.init(categoryRepo!!.first(), true, null, false)
            val dialog = InvoiceItemFormDialog(
                (requireActivity() as AppCompatActivity),
                true,
                null,
                invoiceItemFormViewModel!!
            )
            dialog.show()
            dialog.maximize()
            return@OnEditorActionListener true
        }
        false
    }
    private val ReloadCallback = Runnable {
        Reload(Objects.requireNonNull(QueryEditText().text).toString())
        //
        homeViewModel.filter(QueryEditText().text.toString())
    }
    private val QueryTextChangedListener: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            handler.removeCallbacks(ReloadCallback)
            handler.postDelayed(ReloadCallback, 500)
        }

        override fun afterTextChanged(s: Editable) {
            QueryLayout().error = null
        }
    }
    private val RegisterButtonOnClickListener = View.OnClickListener { v: View? -> Persist(true) }

    @Inject
    fun Inject(
        storeCategoryRepo: IStoreCategoryRepo?,
        storeGroupRepo: IStoreGroupRepo?,
        categoryGroupRepo: ICategoryGroupRepo?,
        GroupRepo: IGroupRepo?,
        storeRepo: IStoreRepo?,
        invoiceRepo: IInvoiceRepo?,
        GroupUnitRepo: IGroupUnitRepo?,
        productRepo: IProductRepo?,
        unitRepo: IUnitRepo?,
        categoryRepo: CategoryRepo?
    ) {
        TheStoreCategoryRepo = storeCategoryRepo
        TheStoreGroupRepo = storeGroupRepo
        TheCategoryGroupRepo = categoryGroupRepo
        TheGroupRepo = GroupRepo
        TheStoreRepo = storeRepo
        TheInvoiceRepo = invoiceRepo
        TheGroupUnitRepo = GroupUnitRepo
        TheProductRepo = productRepo
        TheUnitRepo = unitRepo
        this.categoryRepo = categoryRepo
        init()
    }

    override fun TitleID(): Int {
        return R.string.home_title
    }

    @SuppressLint("NonConstantResourceId")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.ScanMenu -> StartScan()
            R.id.CheckoutMenu -> TryCheckout()
            R.id.DisplayInvoicesMenu -> HostActivity.OnFragmentCalled(null, SabadFragmentEvents.DisplayInvoices)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun StartScan() {
        SabadUtility.BarcodeScanCameraPermission(
            HostActivity,
            {
                val intent = Intent(activity, ScanActivity::class.java)
                startActivityForResult(intent, Requests.Scan.ID)
            },
            false
        )
    }

    private fun TryCheckout() {
        val remainedGroupsParams = GroupParams()
        remainedGroupsParams.Needed = ComparableFilter(true)
        remainedGroupsParams.Picked = ComparableFilter(true)
        remainedGroupsParams.Deleted = ComparableFilter(false)
        val pickedCount = TheGroupRepo!!.Count(remainedGroupsParams)

        if (pickedCount == 0L) {
            FarayanUtility.ShowToastFormatted(activity, "برای ثبت صورتحساب و صندوق، دستکم یکی از نیازمندی‌ها را بردارید")
            return
        }

        DisplayCheckout()
    }

    private fun AutoCheckout() {
        val neededGroupsParams = GroupParams()
        neededGroupsParams.Needed = ComparableFilter(true)
        val neededCount = TheGroupRepo!!.Count(neededGroupsParams)

        val remainedGroupsParams = GroupParams()
        remainedGroupsParams.Needed = ComparableFilter(true)
        remainedGroupsParams.Picked = ComparableFilter(false)
        val remainedCount = TheGroupRepo!!.Count(remainedGroupsParams)

        if (neededCount > 0 && remainedCount == 0L) {
            if (SabadUtility.DisplayCheckoutDialogAutomatically(activity)) {
                if (SabadUtility.CheckoutGuideNeeded(activity)) FarayanUtility.ShowToastFormatted(
                    activity,
                    false,
                    "چون تمام کالاهای فهرست خرید (روشن) را برداشنه‌اید، می‌توانید پایان خرید را بزنید"
                )
                handler.postDelayed({ this.DisplayCheckout() }, 5000)
            }
        }
    }

    private fun DisplayCheckout() {
        val dialog = CheckoutPurchaseDialog(
            CheckoutPurchaseDialog.Input(
                TheStoreCategoryRepo,
                TheStoreGroupRepo,
                TheCategoryGroupRepo,
                TheGroupRepo,
                TheStoreRepo,
                TheInvoiceRepo
            ) { invoiceEntity: InvoiceEntity? ->
                Reload()
                HostActivity.OnFragmentCalled(null, SabadFragmentEvents.DisplayInvoices)
            },
            HostActivity
        )

        dialog.show()
    }

    private fun Persist(toast: Boolean): Boolean {
        try {
            val name = Objects.requireNonNull(QueryEditText().text).toString()
            TheGroupRepo!!.New(name, null, null, true)
            Reload(name)
            QueryLayout().error = null
            if (SabadUtility.NewGroupGuideNeeded(activity)) FarayanUtility.ShowToastFormatted(activity, false, getString(R.string.NewGroupGuide))
            return true
        } catch (e: NewGroupNameNeededException) {
            QueryLayout().error = getString(R.string.NameNeeded)
            if (toast) FarayanUtility.ShowToast(activity, getString(R.string.NameNeeded))
        } catch (e: GroupUniqueNameNeededException) {
            QueryLayout().error = getString(R.string.CategoryUsedName)
            if (toast) FarayanUtility.ShowToast(activity, getString(R.string.CategoryUsedName))
        }
        return false
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.home_menu, menu)
    }

    override fun InitializeLayout() {
        Log.i("flow", "HomeFragment: InitializeLayout")
        homeViewModel = HomeViewModel.Factory.create(HomeViewModel::class.java)
        GroupsRecyclerView().setContent {
            CompositionLocalProvider(LocalLifecycleOwner provides requireActivity(), LocalLayoutDirection provides LayoutDirection.Rtl) {
                val categories = homeViewModel.categories.collectAsState()
                val products = homeViewModel.pickedProducts.collectAsState()
                val units = homeViewModel.pickedUnits.collectAsState()
                val items = homeViewModel.items.collectAsStateWithLifecycle(listOf())
                val removingItem = remember { mutableStateOf<ItemRich?>(null) }
                Column {
                    LazyColumn(
                        modifier = Modifier
                            .fillMaxWidth()
                    ) {
                        items(categories.value, key = { it.id }) { category -> // import androidx.compose.foundation.lazy.items
                            CategoriesCategoryWithItemsComposable(
                                category,
                                items.value,
                                { c, n -> homeViewModel.changeNeeded(c, n) },
                                products.value,
                                units.value,
                                { removingItem.value = it }
                            )
                        }
                    }
                    if (removingItem.value.hasValue) {
                        CategoriesCategoryRemoteItemComposable(
                            removingItem = removingItem.value!!,
                            onConfirmed = { homeViewModel.removeItem(removingItem.value!!.item); removingItem.value = null },
                            onCancelled = { removingItem.value = null }
                        )
                    }
                }
            }
        }
        init()
        QueryEditText().addTextChangedListener(QueryTextChangedListener)
        EditButton().setOnClickListener(EditButtonOnClickListener)
        RegisterButton().setOnClickListener(RegisterButtonOnClickListener)
        QueryEditText().setOnEditorActionListener(QueryEditTextOnEditorActionListener)
    }


    private fun Reload(query: String = Objects.requireNonNull(QueryEditText().text).toString()) {
        var query: String? = query
        query = FarayanUtility.Queryable(query)
        val groupParams = GroupParams()
        groupParams.Deleted = ComparableFilter(false)
        groupParams.QueryableName = TextFilter(query)
        groupParams.Sorts.add(SortConfig(GroupSchema.Needed, SortDirections.Descending))
        groupParams.Sorts.add(SortConfig(GroupSchema.Picked, SortDirections.Ascending))
        groupParams.Sorts.add(SortConfig(GroupSchema.Category, SortDirections.Ascending))
        groupParams.Sorts.add(SortConfig(GroupSchema.PurchasedCount, SortDirections.Descending))
        groupParams.Sorts.add(SortConfig(GroupSchema.Importance, SortDirections.Descending))

        val theAdapter = GroupRecyclerAdapter(
            activity,
            TheGroupRepo!!.All(groupParams),
            { GroupEntity: GroupEntity? ->
                PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient)
                AutoCheckout()
            },
            { GroupEntity: GroupEntity? ->
                PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient)
                AutoCheckout()
            },
            { component: GroupHomeItemComponent ->
                //invoiceItemFormViewModel!!.init(categoryRepo!!.first(), true, null, false)
                val dialog = InvoiceItemFormDialog(
                    (requireActivity() as AppCompatActivity),
                    true,
                    null,
                    invoiceItemFormViewModel!!
                )
                dialog.show()
                dialog.maximize()
            },
            { GroupEntity: GroupEntity? -> Reload() },
            Rial.Coefficients.Rial
        )
        //todo:GroupsRecyclerView().setAdapter(theAdapter)

        groupParams.QueryableName = TextFilter(query, TextMatchModes.Exactly)
        val found = TheGroupRepo!!.Count(groupParams) == 1L
        RegisterButton().visibility = if (found) View.GONE else View.VISIBLE
        EditButton().visibility = if (found) View.VISIBLE else View.GONE
        RegisterButton().isEnabled = FarayanUtility.IsUsable(query)
        PurchaseSummary().ReloadSummary(SabadConstants.TheCoefficient)
    }

    private fun init() {
        if (reloaded) return
        if (TheGroupRepo == null) return
        if (QueryEditText() == null) return
        Reload(Objects.requireNonNull(QueryEditText().text).toString())
        reloaded = true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == Requests.Scan.ID && resultCode == SabadConstants.DataChangedResultCode) {
            Reload()
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun HandleBackPressed(): Boolean {
        if (FarayanUtility.IsUsable(QueryEditText().text)) {
            QueryEditText().setText("")
            return true
        }
        return super.HandleBackPressed()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        //invoiceItemFormViewModel = new ViewModelProvider(requireActivity()).get(InvoiceItemFormViewModel.class);
        invoiceItemFormViewModel = Factory.create(InvoiceItemFormViewModel::class.java)
    }

    enum class Requests(val ID: Int) {
        Scan(1)
    }

}

data class ItemRich(val item: Item, val product: Product, val category: Category)